### PR TITLE
chore(examples): generate examples index

### PR DESCRIPTION
**Pull request by Thomaash**
This generates single index file with (potentialy nested) groups and generated screenshots of examples.

The whole thing is written to be reusable, therefore there should be no problem porting this to Graph3D or Timeline. However it get's almost all of it's information from the examples. All examples have to have the canvas in a container with given id (configurable on CLI). Their titles have to follow this structure:
`Group | [Another Group | [Another Group | …]] Example Title`.

There is also some primitive linting that requires the title to match the heading in the body (not enforced in any way, just a warning).

Result: https://jsfiddle.net/thomaash/fn5mzokq/embedded/result (the links would lead somewhere and the images would't be inlined if this was on GitHub Pages instead of JSFiddle).

PS: I tried to fix this real quick but there may still be some issues with example names as they we're a total mess with a lot of them called Basic Usage or whatever they were originally copy pasted from.

### CODE REVIEW

**Comment by mojoaxel** [comment]
This could be used to improve https://github.com/visjs/vis-graph3d/issues/2

**Comment by Thomaash** [comment]
Maybe we should create vis-dev-util package for this and potentially other reusable tools created in the future so that we can easily share them across projects.

File: examples.template.html [hunk 1]
```diff
@@ -0,0 +1,101 @@
+<html>
+  <head>
+    <meta charset="utf-8" />
+    <title>Vis Network Examples</title>
+
+    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+    <meta name="viewport" content="width=device-width, initial-scale=1" />
+
+    <style>
+      h1 {
```
Comment (mojoaxel) [inline examples.template.html]: I don't think we should any unnecessary headers here. Hopefully at somepoint somebody redesignes this template (we should create an issue for that), but at the moment I would recommend reducing styles in the template to the minimum and let the browsers figure it out how to display it.

File: examples.template.html [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      .example-link .example-image > img {
+        position: absolute;
+        top: 0px;
+        left: 0px;
+        width: 200px;
+        height: 200px;
+
+        border: 1px solid #2b7ce9;
+
+        transition: transform 2s ease 0s, z-index 2s linear 0s;
```
Comment (mojoaxel) [inline examples.template.html]: I think 0.5 is quite enough here :wink:

File: examples.template.html [hunk 3]
```diff
@@ -0,0 +1,101 @@
+<html>
```
Comment (mojoaxel) [inline examples.template.html]: I cont think that the root directory is the best place to put all this. Maybe we introduce a "scripts" folder or something!?

**Comment by mojoaxel** [comment]
It might be a good idea to give every network a "seed" so that they look the same every time they get rendered. This blows up the demo-code a little bit but maybe be could inject this option only while rendering!?

File: generate-example-index.ts [hunk 4]
```diff
@@ -0,0 +1,660 @@
...
+    form.append(
+      cheerio("<input>")
+        .attr("type", "hidden")
+        .attr("name", "title")
+        .attr("value", example.titles.join(" | "))
+    );
+
+    return form;
+  }
+  private _generateCodePen(example: Example): Cheerio {
```
Comment (mojoaxel) [inline generate-example-index.ts]: Maybe we also could add support for jsbin? :pray: 
https://jsbin.com/help/api/

File: examples.template.html (hunk 1 above)
Comment (mojoaxel) [reply examples.template.html]: On mobile the fonts are also WAY to big

Comment (Thomaash) [reply examples.template.html]: The template is only for the index file(s). The examples are kept intact, no harm done. They have optional meta headers but these can be avoided by rewriting the example (unless it has some special needs which is the reason why they exists but that should be rare).

File: generate-example-index.ts (hunk 4 above)
Comment (Thomaash) [reply generate-example-index.ts]: JS Bin seems to have API only for pro JS Bin account (which I don't have).

Comment (Thomaash) [reply generate-example-index.ts]: Also wouldn't committing the token to a public repo be security risk? Couldn't it be abused?

File: generate-example-index.ts [hunk 5]
```diff
@@ -0,0 +1,660 @@
...
+    return {
+      code: {
+        css,
+        html,
+        js
+      },
+      resources
+    };
+  }
+  private _generateJSFiddle(example: Example): Cheerio {
```
Comment (mojoaxel) [inline generate-example-index.ts]: For me jsfiddle always runs into a timeout.

Comment (Thomaash) [reply generate-example-index.ts]: It's not just you! jsfiddle.net is down: http://isup.me/jsfiddle.net .

File: examples.template.html (hunk 3 above)
Comment (Thomaash) [reply examples.template.html]: Done.

File: generate-example-index.ts (hunk 4 above)
Comment (mojoaxel) [reply generate-example-index.ts]: > JS Bin seems to have API only for pro JS Bin account (which I don't have).

I love jsbin and I'm using it for almost a decade I think. 
We could talk to @remy maybe he feels supportive and gives us access for free. I don't think we create any relative traffic...

File: package.json [hunk 6]
```diff
@@ -36,6 +36,7 @@
     "build": "npm run types:network && npm run js:network && npm run images:network",
     "watch": "npm run images:network && rollup --watch --config rollup.config.js",
     "watch-dev": "npm run watch-dev",
+    "generate-examples-index": "node -r ./babel.mocha.js generate-example-index.ts -c mynetwork -w https://visjs.github.io/vis-network -lis",
```
Comment (mojoaxel) [inline package.json]: We should think about calling this from within the CI. This could be done together with #95

Comment (Thomaash) [reply package.json]: Yeah, when #95 is merged I'll add this to it.

Comment (Thomaash) [reply package.json]: Or the other way around if this is merged first.

File: examples.template.html (hunk 1 above)
Comment (mojoaxel) [reply examples.template.html]: I'm talking about the overview index file.

Comment (Thomaash) [reply examples.template.html]: Why? Does it matter? It's just an index file. Nobody's going to read this anyway.

File: examples.template.html (hunk 2 above)
Comment (Thomaash) [reply examples.template.html]: Done. It really feels better.

File: examples.template.html (hunk 1 above)
Comment (Thomaash) [reply examples.template.html]: I din't check it on mobile. That's crazy!

Comment (Thomaash) [reply examples.template.html]: It should be reasonable now.

**Comment by mojoaxel** [comment]
> It might be a good idea to give every network a "seed" so that they look the same every time they get rendered. This blows up the demo-code a little bit but maybe be could inject this option only while rendering!?

@Thomaash What do you think about that?
The option I have in mind is the `randomSeed` in the [layout config](https://visjs.github.io/vis-network/docs/network/layout.html).

File: examples/network/layout/hierarchicalLayoutDynamicChanges.html [hunk 7]
```diff
@@ -0,0 +1,132 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <title>Vis Network | Layouts | Dynamic Hierarchical Layout Changes</title>
```
Comment (mojoaxel) [inline examples/network/layout/hierarchicalLayoutDynamicChanges.html]: I'm not sure if this PR is the best place to introduce a completely new example, but I don't mind!

File: generate-example-index.ts (hunk 4 above)
Comment (mojoaxel) [reply generate-example-index.ts]: Ok, let's do this in a later step - if possible...

File: examples/network/layout/hierarchicalLayoutDynamicChanges.html (hunk 7 above)
Comment (Thomaash) [reply examples/network/layout/hierarchicalLayoutDynamicChanges.html]: This isn't exactly new. It slipped here from #115.

**Comment by Thomaash** [comment]
> > It might be a good idea to give every network a "seed" so that they look the same every time they get rendered. This blows up the demo-code a little bit but maybe be could inject this option only while rendering!?
> 
> @Thomaash What do you think about that?
> The option I have in mind is the `randomSeed` in the [layout config](https://visjs.github.io/vis-network/docs/network/layout.html).

This was a bit tricky but done. I'll have a look at exampleUtil now because it uses Math.random() and screws things up. This should be an easy fix.

**Comment by Thomaash** [comment]
No problem.

**Pull request merged**